apps/src/gamelab/GameLab.js

```diff
@@ -557,6 +581,43 @@ GameLab.prototype.reset = function (ignore) {
   }
 };
 
+GameLab.prototype.runSetupCode = _.debounce(function () {
+  if (getStore().getState().runState.isRunning) {
+    return;
+  }
+  const visualizationCanvas = document.getElementById('defaultCanvas0');
+  if (visualizationCanvas) {
+    // Cover the playspace with a copy of itself while rerunning the setup code
+    // to avoid flickering
+    const duplicateCanvas = document.createElement('canvas');
+    const visualizationRect = visualizationCanvas.getBoundingClientRect();
+    duplicateCanvas.width = visualizationRect.width;
+    duplicateCanvas.height = visualizationRect.height;
+    Object.assign(duplicateCanvas.style, {
+      position: 'absolute',
+      top: visualizationRect.top + 'px',
+      left: visualizationRect.left + 'px',
+      zIndex: 1,
+    });
+    const duplicateCtx = duplicateCanvas.getContext('2d');
+    duplicateCtx.drawImage(visualizationCanvas,
+      0, 0,
+      visualizationRect.width, visualizationRect.height
+    );
+
+    document.body.appendChild(duplicateCanvas);
+    setTimeout(() => document.body.removeChild(duplicateCanvas), 0);
```

Comment:
I can't decide if this feels like a total hack, or an elegant solution to the redraw problem 😄.  Ideally p5 would take an upstream change to give us more hooks, so we could get flicker-free-redraws without the duplicate canvas.

Another option: can we keep p5 running, clear the canvas, then redraw to avoid the flicker?

Reply:
This is sort of awesome hacky double-buffering :grin: and I'd definitely suggest we use p5 for this ([see `createGraphics`](https://p5js.org/examples/structure-create-graphics.html)) if it's easy to keep an instance around, but I'm not sure it is.  I'm fine with this approach but would suggest several adjustments.

- First, there are some good comments here but for something like this I'd appreciate a larger comment (maybe on the function) describing the high-level approach.
- Even though you've throttled this method, it's going to run a _lot_ so doing some easy optimizations now is probably worth it.
  - Avoid creating elements in this method; let's make one duplicateCanvas and one loadingElement, and reuse them.
  - Even cutting DOM queries - we can probably safely cache a reference to the visualizationCanvas.